### ISSUE TITLE
Consider cloning @options when OmniAuth::Strategy is dup'd

### ISSUE THREAD

**Issue by md5** [omniauth/omniauth#823]
I ran into an issue yesterday with `omniauth-saml` where the `options` data structure was being mutated in `request_phase` in a way that making state bleed across requests (see https://github.com/PracticallyGreen/omniauth-saml/issues/54).

After some initial confusion on my part about the middleware lifecycle and looking through a bunch of code, I found that OmniAuth is doing `dup` before it calls the strategy methods (cf. https://github.com/intridea/omniauth/blob/c3e3aa5bbfb589db1ebed40324c7eea2e3153f8c/lib/omniauth/strategy.rb#L164).

The issue I want to raise here is whether or not it makes sense for `OmniAuth::Strategy` to implement `initialize_copy` like so:

``` ruby
def initialize_copy(orig)
  super
  @options = @options.clone
end
```

This would make it safe for strategy subclasses to mutate `options` in `request_phase`, etc. It might actually need to be `@options.deep_dup` instead of `@options.clone`, but you get the idea.

Does this make sense as a PR for OmniAuth or should I stick with making a PR for `omniauth-saml` that avoids mutating the shared instance of `OmniAuth::Strategy::Options`? If this change to OmniAuth isn't desirable, perhaps it should be documented that it isn't safe for subclasses of `OmniAuth::Strategy` to mutate `options` without making a local copy first.

**Comment by md5** [comment]
If the documentation route is preferable, I can add some text to the [Strategy Contribution Guide](https://github.com/intridea/omniauth/wiki/Strategy-Contribution-Guide) on the wiki.

**Comment by md5** [comment]
I ended up submitting a PR to `omniauth-saml`: https://github.com/PracticallyGreen/omniauth-saml/pull/56

If there's a favorable response to my suggestion here, I can close that one and make a similar PR for `omniauth` itself.

**Comment by md5** [comment]
Bumping this issue to see if anyone has any opinions about implementing a generalized fix for this issue in `omniauth` itself instead of expecting individual strategies to do it

**Comment by md5** [comment]
This was fixed in #829
